fix: BottomNavigationBarItem colors

talamaska opened the pull request:
add: _effectiveIconTheme function
add: logic for setting the effective selected/unselected icon theme 
change: color tweens to use the effective label style color if not null
change: use the new effective selected/unselected Icon theme when generating _BottomNavigationTile
add: tests to capture the issues with the label style color, matching the docs

fix: BottomNavigationBarItem title's color is not overriden neither by selectedLabelStyle nor unselectedLabelStyle


#64358

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

talamaska committed:
fix: issue 64358

add: _effectiveIconTheme function
add: logic for setting the effective selected/unselected icon theme 
change: color tweens to use the effective label style color if not null
change: use the new effective selected/unselected Icon theme when generating _BottomNavigationTile
add: tests to capture the issues with the label style color, matching the docs

talamaska commented:
I would love to get some feedback on the PR. I'm prepared to make any additional changes to meet the requirements.

darrenaustin commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -889,6 +889,12 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
   }
 
+  static IconThemeData _effectiveIconTheme(IconThemeData? iconTheme, Color? itemColor) {
+    iconTheme ??= const IconThemeData();
+    // Prefer the font size on textStyle if present.
+    return itemColor != null && iconTheme == null ? iconTheme.copyWith(color: itemColor) : iconTheme;
```
I think this can be simplified to just: 

```dart
    return (iconTheme ?? const IconThemeData()).copyWith(color: itemColor);
```

darrenaustin commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,207 @@ void main() {
+    expect(unselectedIcon.color, equals(unselectedColor));
+  });
+
+  testWidgets('Fixed BottomNavigationBar label style color should override itemColor', (WidgetTester tester) async {
```
```suggestion
  testWidgets('label style color should override itemColor for BottomNavigationBarType.fixed'', (WidgetTester tester) async {
```

darrenaustin commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,207 @@ void main() {
+    const Color selectedLabelColor = Color(0xFFFF9900);
+    const Color unselectedLabelColor = Color(0xFF92F74E);
+
+
```
Small nit: no need for the extra blank line.

darrenaustin commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,207 @@ void main() {
+    expect(unselectedIcon.color, equals(unselectedLabelColor));
+  });
+
+  testWidgets('Shifting BottomNavigationBar label style color should override itemColor', (WidgetTester tester) async {
```
```suggestion
  testWidgets('label style color should override itemColor for BottomNavigationBarType.shifting', (WidgetTester tester) async {
```

darrenaustin commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,207 @@ void main() {
+    const Color selectedLabelColor = Color(0xFFFF9900);
+    const Color unselectedLabelColor = Color(0xFF92F74E);
+
+
```
Nit: extra blank line.

darrenaustin commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,207 @@ void main() {
+    expect(unselectedIcon.color, equals(unselectedLabelColor));
+  });
+
+  testWidgets('Fixed BottomNavigationBar label style color should not override iconTheme color', (WidgetTester tester) async {
```
```suggestion
  testWidgets('iconTheme color should override label style color for BottomNavigationBarType.fixed', (WidgetTester tester) async {
```

darrenaustin commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,207 @@ void main() {
+    const TextStyle selectedTextStyle = TextStyle(fontSize: 18.0, color: selectedLabelColor);
+    const TextStyle unselectedTextStyle = TextStyle(fontSize: 18.0, color: unselectedLabelColor);
+
+
```
Nit: extra blank line.

darrenaustin commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,207 @@ void main() {
+    expect(unselectedIcon.color, equals(unselectedIconthThemeColor));
+  });
+
+  testWidgets('Shifting BottomNavigationBar label style color should not override iconTheme color', (WidgetTester tester) async {
```
```suggestion
  testWidgets('iconTheme color should override label style color for BottomNavigationBarType.shifted', (WidgetTester tester) async {
```

darrenaustin commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,207 @@ void main() {
+    const TextStyle selectedTextStyle = TextStyle(fontSize: 18.0, color: selectedLabelColor);
+    const TextStyle unselectedTextStyle = TextStyle(fontSize: 18.0, color: unselectedLabelColor);
+
+
```
Nit: extra blank line.

darrenaustin reviewed (changes requested):
Thanks for the contribution! It appears this is a bit of a confusing API, so thanks for cleaning it up. In addition to the comments and suggestions below, can we add something to the class doc comment about the `IconThemeData` taking precedence over `iconSize` and item colors?

darrenaustin commented:
It also appears that the analyzer is complaining about trailing spaces on the ends of some lines. (See the `analyze-linux` check's details)

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/96664

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 5.
View them at https://flutter-gold.skia.org/cl/github/96664

skia-gold commented:
Gold has detected about 8 new digest(s) on patchset 7.
View them at https://flutter-gold.skia.org/cl/github/96664

skia-gold commented:
Gold has detected about 6 new digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/96664

talamaska commented:
@darrenaustin fixed everything.
Added more docs and removed the trailing spaces
Seems like there is a problem with the CLA because I accepted your commit suggestions. Is that resolvable or I should reset and apply the changes myself.

skia-gold commented:
Gold has detected about 18 new digest(s) on patchset 10.
View them at https://flutter-gold.skia.org/cl/github/96664

talamaska committed:
PR review changes

change: rewording in test cases
remove: trailing spaces
remove blank lines
add: docs for IconThemeData's
fix: typos

talamaska commented:
reset and re-applied changes to avoid CLA issues and not waste more time on this.

darrenaustin commented:
@talamaska thanks for the update. It looks like the new tests are failing however. From the logs for one of the failing checks:

```
00:35 +626: /b/s/w/ir/x/w/flutter/packages/flutter/test/material/bottom_navigation_bar_test.dart: label style color should override itemColor for BottomNavigationBarType.shifting                     
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: Color:<Color(0xffff9900)>
  Actual: Color:<Color(0xff0004ff)>
```

Can you take a look? Thx.

talamaska commented:
@darrenaustin it seems your suggestion for shortening is not working as expected. 
```dart
 return (iconTheme ?? const IconThemeData()).copyWith(color: itemColor);
 ```
If i return the previous way of writing
```dart
iconTheme ??= const IconThemeData()
// Prefer the iconTheme over itemColor
return itemColor != null && iconTheme == null ? iconTheme.copyWith(color: itemColor) : iconTheme;

```

talamaska committed:
fix: _effectiveIconTheme logic

talamaska committed:
fix: inline docs for _effectiveIconTheme

add: more tests
change: _effectiveIconTheme to account for textStyle

talamaska commented:
I think it is fixed now.
Added more docs as well.

talamaska committed:
remove: trailing space in doc comments

talamaska commented:
@darrenaustin is there something else that is blocking this PR?

talamaska commented:
@darrenaustin, hi there. I wonder, can I get a review again? It's been quite some time since I implemented the requested changes. I would really like this to be merged before starting other efforts to eventual material 3, because I'm afraid the issues will end up in the new version as well. I have pinged in discord as well. If it is more convenient we can chat there.

darrenaustin commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -889,6 +895,13 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
   }
 
+  static IconThemeData _effectiveIconTheme(IconThemeData? iconTheme, Color? itemColor, TextStyle? textStyle) {
+    final IconThemeData effectiveIconTheme = iconTheme ?? const IconThemeData();
+    // Prefer the iconTheme over itemColor and textStyle color if present.
+    // If iconTheme is not set, prefer textStyle color over itemColor
+    return iconTheme == null ? effectiveIconTheme.copyWith(color: textStyle?.color ?? itemColor) : effectiveIconTheme;
```
If I am not mistaken, can't this be simplified to just:

```dart
return iconTheme ?? IconThemeData(color: textStyle?.color ?? itemColor);
```

darrenaustin commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -907,6 +920,20 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
+      _effectiveIconTheme(
+        widget.selectedIconTheme ?? bottomTheme.selectedIconTheme,
+        widget.selectedItemColor,
+        widget.selectedLabelStyle,
```
Why are we using  the label style's color for the icon? It describes the text color, not the icon color.

talamaska replied:
@darrenaustin  You are correct. I don't know how I got confused. Will make changes and look again at the color tween.

talamaska replied:
@darrenaustin Fixed by removing the dependency on the text style, but I was thinking more about this. 
Shouldn't i make the effectiveIconTheme depend on bottomTheme and themeColor, so it ends up like this
```dart
    final IconThemeData effectiveSelectedIconTheme =
      _effectiveIconTheme(
        widget.selectedIconTheme ?? bottomTheme.selectedIconTheme,
        widget.selectedItemColor 
        ?? bottomTheme.selectedItemColor 
        ?? themeColor
      );
 ```
 
Unfortunately this also breaks the tests for the shifting backgroundColor

darrenaustin commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -907,6 +920,20 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
+      _effectiveIconTheme(
+        widget.unselectedIconTheme ?? bottomTheme.unselectedIconTheme,
+        widget.unselectedItemColor,
+        widget.unselectedLabelStyle,
```
Same comment here: why are we using the label style's color for the icon? It describes the text color, not the icon color.

talamaska replied:
@darrenaustin Same logic here
```dart
    final IconThemeData effectiveUnselectedIconTheme =
      _effectiveIconTheme(
        widget.unselectedIconTheme ?? bottomTheme.unselectedIconTheme,
        widget.unselectedItemColor 
        ?? bottomTheme.unselectedItemColor 
        ?? themeData.unselectedWidgetColor
      );
```

darrenaustin commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -921,21 +948,25 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
       case BottomNavigationBarType.fixed:
         colorTween = ColorTween(
-          begin: widget.unselectedItemColor
+          begin: effectiveUnselectedLabelStyle.color
+            ?? widget.unselectedItemColor
             ?? bottomTheme.unselectedItemColor
             ?? themeData.unselectedWidgetColor,
-          end: widget.selectedItemColor
+          end: effectiveSelectedLabelStyle.color
+            ?? widget.selectedItemColor
             ?? bottomTheme.selectedItemColor
             ?? widget.fixedColor
             ?? themeColor,
```
This `ColorTween` is a bit odd in that it is used for both the icon and label colors, even though they can be different colors. I know this was already in place before your fix, but this just seems to conflate the two even more. Perhaps we should break these out into a tween for the icon and a tween for the text.

talamaska replied:
@darrenaustin So, if i make the previous changes without changing this color tween, all tests are successful, as soon I add a separate tween for the icon colors some golden tests fail.

darrenaustin reviewed (changes requested):
@talamaska, thanks so much for your patience.  Sorry I haven't been able to get to this sooner.

Taking a closer look, I am confused why we are conflating the icon and label colors (other than through the `(un)selectedItemColor`). This conflation in the tweens was there before your fix, but it seems wrong to keep mixing the two. Perhaps I am missing something though.

Also, I fear this may be a breaking change, as I tried this patch against internal Google tests and it failed several golden tests. There appear to be usages of this API that depend on the current behaviour. We might be able to either fix the tests or ensure they don't break, but first I want to see if we can resolve the icon vs. label color issues.

talamaska committed:
fix: icon color should not depend on label style

fix: tests for icon color

talamaska committed:
Merge branch 'master' into fix/64358-wrong-colors-in-bottom-navigation-bar

talamaska commented:
I'm also kinda confused why Shifting bottom bar is so broken and should I try to fix that as well? 
![image](https://user-images.githubusercontent.com/299973/165368639-36fe1ae4-4262-4177-bd25-e934629bee19.png)
vs this in fixed
![image](https://user-images.githubusercontent.com/299973/165368694-d7804fa4-f91d-4976-8144-771250026752.png)

In the end if I'm going to implement the above suggested selected and unselected effectiveIconTheme 
```dart
final IconThemeData effectiveSelectedIconTheme =
      _effectiveIconTheme(
        widget.selectedIconTheme ?? bottomTheme.selectedIconTheme,
        widget.selectedItemColor 
        ?? bottomTheme.selectedItemColor 
        ?? themeColor
      );
      
final IconThemeData effectiveUnselectedIconTheme =
      _effectiveIconTheme(
        widget.unselectedIconTheme ?? bottomTheme.unselectedIconTheme,
        widget.unselectedItemColor 
        ?? bottomTheme.unselectedItemColor 
        ?? themeData.unselectedWidgetColor
      );
```
and breaking that golden test, it woule make sens to implement the colorTween separation as well

talamaska commented:
@darrenaustin in the end, would it make send to drop all those fixes if there is a material 3 rewrite going? Or is the new material 3 navigation bar will be a new widget?

TahaTesser commented:
I saw https://github.com/flutter/flutter/issues/96480 this morning and filed a PR for it, which addresses `selectedLabelStyle` and `unselectedLabelStyle`. 

Now I found this PR and https://github.com/flutter/flutter/issues/64358



Just closed new issue https://github.com/flutter/flutter/issues/96480 for a duplicate of the issue linked here https://github.com/flutter/flutter/issues/64358

@darrenaustin

> fix: BottomNavigationBarItem title's color is not overriden neither by selectedLabelStyle nor unselectedLabelStyle

I think my PR addressed the issue matching the Flutter docs 



> fix: BottomNavigationBarItem icon's color is not overriden neither by selectedLabelStyle nor unselectedLabelStyle

`selectedLabelStyle` or `unselectedLabelStyle` shouldn't update BottomNavigationBarItem icon's color.

TahaTesser commented:
I saw https://github.com/flutter/flutter/issues/96480 this morning and filed a PR for it, which addresses `selectedLabelStyle` and `unselectedLabelStyle`. 

Now I found this PR and https://github.com/flutter/flutter/issues/64358



Just closed new issue https://github.com/flutter/flutter/issues/96480 for a duplicate of the issue linked here https://github.com/flutter/flutter/issues/64358

@darrenaustin

> fix: BottomNavigationBarItem title's color is not overriden neither by selectedLabelStyle nor unselectedLabelStyle

I think my PR addressed the issue matching the Flutter docs 



> fix: BottomNavigationBarItem icon's color is not overriden neither by selectedLabelStyle nor unselectedLabelStyle

`selectedLabelStyle` or `unselectedLabelStyle` shouldn't update BottomNavigationBarItem icon's color.

talamaska committed:
Merge branch 'master' into fix/64358-wrong-colors-in-bottom-navigation-bar

talamaska commented:
@darrenaustin @TahaTesser if my PR is not needed anymore feel free to close. It's a shame that I was waiting for review months and @TahaTesser PR was reviewed and approved in a matter of hours. I understand that my PR was not perfect, but anyways. 
I'm happy that's resolved. Do you think the shifting bottom bar colors should be fixed in a separate PR and update it's goldens?

TahaTesser commented:
@talamaska 
My PR had to be reverted https://github.com/flutter/flutter/pull/102756, I would suggest keeping this PR open and making sure to address the issues related to revert in your PR.

Hixie commented:
What is the status of this PR? Did we end up fixing this issue in the other PR?

@talamaska sorry for the delays in dealing with this PR. You are welcome to come bug us on our Discord server (link in the contributing guidelines) if you would like to ask for reviews when we drop the ball. :-)

talamaska commented:
@Hixie Hi, well, it seems the PR is a no go, because it is breaking some golden tests. Which on my opinion just rely on a wrong behavior and in addition the reported issues are not included in the existing tests. I didn't get any feedback on the question about the shifting bottom bar styles or any other question in my comment to @darrenaustin. In the other PR https://github.com/flutter/flutter/pull/102756 that was reverted I got an explanation that this is a breaking change and there is a different process of implementing those - with temporary flags or something else that can gracefully help devs migrate to the new behavior. So I'm not sure what to do here.

talamaska committed:
fix: icon color should not depend on text style color
add: a separate tween of the icons
fix: tests

talamaska committed:
Merge branch 'master' into fix/64358-wrong-colors-in-bottom-navigation-bar

talamaska committed:
remove: trailing spaces

Hixie commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -154,6 +154,12 @@ class BottomNavigationBar extends StatefulWidget {
   /// [selectedIconTheme] and [unselectedIconTheme], and both
   /// [IconThemeData.color] and [IconThemeData.size] must be set.
   ///
+  /// If set, [IconThemeData]s take precedence over [iconSize],
+  /// [selectedItemColor] and [unselectedItemColor].
+  ///
+  /// If set, [IconThemeData]s take precedence over [selectedItemColor] and [unselectedItemColor],
+  /// for the icons only
```
This might be clearer as:

```suggestion
  /// [IconThemeData]s take precedence over [iconSize], [selectedItemColor]
  /// and [unselectedItemColor] for icons.
```

...but I'm not quite sure what "if set" means so it's hard for me to say (I'm not super familiar with this class). Does it specifically refer to selectedIconTheme/unselectedIconTheme, or is it referring to ambient themes?

talamaska replied:
@Hixie It refers to selectedIconTheme/unselectedIconTheme. But you are correct in your assumed suggestion.

Hixie commented on packages/flutter/lib/src/material/bottom_navigation_bar.dart:
```diff
@@ -928,35 +953,76 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
-      case Brightness.light:
-        themeColor = themeData.colorScheme.primary;
+    final IconThemeData effectiveSelectedIconTheme =
+      _effectiveIconTheme(
+        widget.selectedIconTheme ?? bottomTheme.selectedIconTheme,
+        widget.selectedItemColor
+        ?? bottomTheme.selectedItemColor
+        ?? themeColor
+      );
```
(same applies in many other places; for readability, we should be consistent about how we do this throughout)

Hixie commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -533,9 +540,286 @@ void main() {
+    const IconThemeData selectedIconTheme = IconThemeData(size: 20, color: selectedIconThemeColor);
+    const IconThemeData unselectedIconTheme = IconThemeData(size: 18, color: unselectedIconThemeColor);
+
+
```
extraneous blank line

Hixie commented:
If it's breaking some customer tests then yeah, generally that means we end up putting a flag that flips between the old and new behaviours for a while, so that we can migrate the code. That's relatively straight forward; you just have a static boolean variable `BottomNavigationBar.useLegacyColorScheme` that defaults to `true`, and you implement your logic only when it's false. Then we land that, flip the flag in any customer code, make the tests pass, then we flip the default for the flag and remove the old behaviour, land that, remove the flag flip in the customer code, and then finally remove the flag. Ideally if we do this we'd also publish a migration guide, which for this kind of change would be really trivial, just telling people there's nothing to change, all we've done is make the colors better, and pointing them to the properties they can use to control the colors if they prefer the old styles.

Hixie commented:
I couldn't figure out what the shifting label issue was, can you elaborate on this? Where do the screenshots come from?

For the migration guide the documentation for doing that is in the template: https://github.com/flutter/website/blob/main/src/release/breaking-changes/template.md 

More information about breaking changes is at: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

talamaska committed:
fix: issues from PR review

change: tests
fix: docs
fix: shifting bottom navigation bar colors
add: legacy flag
add: conditions for loading different tweens and themes depending on the legacy flag value.

talamaska commented:
@Hixie will check the info about the breaking changes. For now the flag is implemented and all the conditional logic is there, so with the flag true, no tests are breaking and tests are testing with the flag equal to false.

talamaska committed:
remove: trailing space

talamaska commented:
Could you please elaborate on squashing and rebasing. I have like 3 merges from master thought the life of this branch. Do I need to revert from the beginning, make my changes, then rebase master into my branch? I'm sorry if this is a trivial question, but I have never rebased before and I see I can squash commits if they include a merge in the middle. Would it be easier to just drop this PR, make a new branch from the current master, add my changes and submit a new PR?

Hixie commented:
Easiest way is probably to just create a new branch, then cherry-pick your commits onto it. Alternatively you might be able to do it using `git rebase -i` where you just select your commits and skip the merge commits (I'm not sure exactly how that works since I don't normally try to use merges). In general for Flutter we recommend using rebases, see also https://github.com/flutter/flutter/wiki/Tree-hygiene#using-git

talamaska commented:
@darrenaustin @Hixie closing this PR in favor of a new one #107924
